pandas/core/frame.py

```diff
@@ -3122,6 +3122,8 @@ def _setitem_frame(self, key, value):
 
         self._check_inplace_setting(value)
         self._check_setitem_copy()
+        if not self._indexed_same(key):
+            key = key.reindex_like(self).fillna(False)
```

Comment:
I did this because inside _where we're filling with False and so we end up incorrectly setting values that aren't selected by key (previously the correct fill_value seemingly was only used "by accident" since we had inplace=True and also fill_value = bool(inplace)).

Reply:
instead of this ok to add an arg to _where to do the inverting. *all* logic associated with the reindexing / filling must live there.